src/Storages/PostgreSQL/StorageMaterializedPostgreSQL.cpp

```diff
@@ -362,53 +363,118 @@ ASTPtr StorageMaterializedPostgreSQL::getColumnDeclaration(const DataTypePtr & d
         create_table_query->uuid = table_id.uuid;
 
+    auto storage = std::make_shared<ASTStorage>();
+    storage->set(storage->engine, makeASTFunction("ReplacingMergeTree", std::make_shared<ASTIdentifier>("_version")));
+
     auto columns_declare_list = std::make_shared<ASTColumns>();
-    auto columns_expression_list = std::make_shared<ASTExpressionList>();
     auto order_by_expression = std::make_shared<ASTFunction>();
 
     auto metadata_snapshot = getInMemoryMetadataPtr();
-    const auto & columns = metadata_snapshot->getColumns();
+
+    ConstraintsDescription constraints;
     NamesAndTypesList ordinary_columns_and_types;
 
-    if (!is_materialized_postgresql_database)
-    {
-        ordinary_columns_and_types = columns.getOrdinary();
-    }
-    else
+    if (is_materialized_postgresql_database)
     {
-        if (!table_structure)
+        if (!table_structure && !table_override)
         {
-            throw Exception(ErrorCodes::LOGICAL_ERROR,
-                    "No table structure returned for table {}.{}", table_id.database_name, table_id.table_name);
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "No table structure returned for table {}.{}",
+                            table_id.database_name, table_id.table_name);
         }
 
-        if (!table_structure->columns)
+        if (!table_structure->columns && (!table_override || !table_override->columns))
         {
-            throw Exception(ErrorCodes::LOGICAL_ERROR,
-                    "No columns returned for table {}.{}", table_id.database_name, table_id.table_name);
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "No columns returned for table {}.{}",
+                            table_id.database_name, table_id.table_name);
         }
 
-        ordinary_columns_and_types = *table_structure->columns;
+        bool has_order_by_override = table_override && table_override->storage && table_override->storage->order_by;
+        if (has_order_by_override && !table_structure->replica_identity_columns)
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                            "Having PRIMARY KEY OVERRIDE is allowed only if there is "
+                            "replica identity index for PostgreSQL table. (table {}.{})",
+                            table_id.database_name, table_id.table_name);
+        }
 
         if (!table_structure->primary_key_columns && !table_structure->replica_identity_columns)
         {
             throw Exception(ErrorCodes::BAD_ARGUMENTS,
-                    "Table {}.{} has no primary key and no replica identity index", table_id.database_name, table_id.table_name);
+                            "Table {}.{} has no primary key and no replica identity index",
+                            table_id.database_name, table_id.table_name);
         }
 
+        if (table_override && table_override->columns)
+        {
+            table_override->applyToCreateTableQuery(create_table_query.get());
```

Comment:
It's already applied [here](https://github.com/ClickHouse/ClickHouse/blob/47f549166d5c1b7932067d145cf80a2b469c3901/src/Storages/PostgreSQL/StorageMaterializedPostgreSQL.cpp#L389-L390), why do we apply it twice?